aip/auth/4120.md

```diff
@@ -0,0 +1,96 @@
+The token exchange for Service Accounts is not a supported flow outside the GDU, and services must support SSJ with scopes. Therefore authentication libraries **must** use the self-signed JWT flow when authenticating outside the GDU using Service Account credentials. If the `useJwtWithScopes` parameter exists (whatever form it takes in specific languages), it **must** be ignored outside the GDU, and its documentation **must** be updated to reflect that.
+
+### GCE credentials retrieve the universe domain from the metadata server
+
+> This requirement applies **only** to the VM metadata flow.
+
+GCE credentials (also known as VM Metadata credentials) directly retrieves an access token from the Metadata Server (MDS). The access token has no universe information. The authentication libraries **must** modify the VM metadata flow to make a second request to retrieve the universe_domain from the following URI: `http://metadata.google.internal/computeMetadata/v1/universe/universe_domain`.
+
+Once the `universe_domain` is retrieved, the authentication libraries **must** cache it to avoid making unnecessary MDS requests.
+
+Failure modes when retrieving the universe_domain from MDS:
+
+| Error when retrieving the universe_domain from MDS | Action |
+| -------------------------------------------------- | ------ |
+| 404 | The authentication libraries **must** set the `universe_domain` to the default value `"googleapis.com"`. |
+| Timeout | The authentication libraries **must not** fall back to the `"googleapis.com"` default value. A retry mechanism is recommended. The authentication libraries **should** do the same action as when retrieving the access token times out. |
+| Any other error | The authentication libraries **must not** fall back to the `"googleapis.com"` default value. The authentication libraries **should** do the same action as when the same error happens when retrieving the access token. |
+
+Additional recommendations:
+
+ - GCE credentials **should** retrieve the universe domain in a 'lazy' way, when the end-user requests the `universe_domain` via the property/get method. They **should not** do it during the credentials initialization.
+
+### Other requirements
+
+#### Authentication libraries **must not** use hardcoded non-constant information in the authentication flows.
+
+> This requirement applies **only** to the flows which are supported outside the GDU.
+
+The code for the authentication flows **must not** contain hardcoded endpoints, endpoint-alikes or other authentication information unless this information is constant across all flows for all universes. Any information that is possible to read from the credentials file should be read from the credentials file. For example, `scopes` contain the "googleapis.com" domain but are still valid outside the GDU because their values are consistent across all universes, whereas the STS endpoint "sts.googleapis.com" is not allowed because it is different for every universe.
```

Comment:
Is there a reference for what changes outside and what doesn't that we can refer to?

Reply:
There is no canonical list that I am aware of. I believe this is because "scopes" is currently the only value of `googleapis.com` that is allowed - everything else should change based on the universe.